apps/api/src/app/workflows-v2/usecases/upsert-workflow/upsert-workflow.usecase.ts

```diff
@@ -0,0 +1,306 @@
+      description: workflowDto.description || '',
+      tags: workflowDto.tags || [],
+      critical: false,
+    };
+  }
+
+  private async getWorkflowIfUpdateAndExist(upsertCommand: UpsertWorkflowCommand) {
+    if (upsertCommand.workflowDatabaseIdForUpdate) {
+      return await this.notificationTemplateRepository.findByIdQuery({
+        id: upsertCommand.workflowDatabaseIdForUpdate,
+        environmentId: upsertCommand.user.environmentId,
+      });
+    }
+  }
+
+  private async workflowExistByExternalId(upsertCommand: UpsertWorkflowCommand) {
+    const { environmentId } = upsertCommand.user;
+    const workflowByDbId = await this.notificationTemplateRepository.findByTriggerIdentifier(
+      environmentId,
+      upsertCommand.workflowDto.name
+    );
+
+    return !!workflowByDbId;
+  }
+
+  private convertCreateToUpdateCommand(
+    command: UpsertWorkflowCommand,
+    existingWorkflow: NotificationTemplateEntity
+  ): UpdateWorkflowCommand {
+    const { workflowDto } = command;
+    const { user } = command;
+
+    return {
+      id: existingWorkflow._id,
+      environmentId: user.environmentId,
+      organizationId: user.organizationId,
+      userId: user._id,
+      name: command.workflowDto.name,
+      steps: this.mapSteps(workflowDto.steps, existingWorkflow),
+      rawData: workflowDto,
+      type: WorkflowTypeEnum.BRIDGE,
+      description: workflowDto.description,
+      tags: workflowDto.tags,
+      active: workflowDto.active ?? true,
+    };
+  }
+
+  private mapSteps(
+    commandWorkflowSteps: Array<StepCreateDto | StepUpdateDto>,
+    persistedWorkflow?: NotificationTemplateEntity | undefined
+  ): NotificationStep[] {
+    const steps: NotificationStep[] = commandWorkflowSteps.map((step) => {
+      return this.mapSingleStep(persistedWorkflow, step);
+    });
+
+    return steps;
+  }
+
+  private mapSingleStep(
+    persistedWorkflow: NotificationTemplateEntity | undefined,
+    step: StepDto | (StepDto & { stepUuid: string })
+  ): NotificationStep {
+    const stepEntityToReturn = this.buildBaseStepEntity(step);
+    const foundPersistedStep = this.getPersistedStepIfFound(persistedWorkflow, step);
+    if (foundPersistedStep) {
+      return {
+        ...stepEntityToReturn,
+        _id: foundPersistedStep._templateId,
+        _templateId: foundPersistedStep._templateId,
+        template: { ...stepEntityToReturn.template, _id: foundPersistedStep._templateId },
+      };
+    }
+
+    return stepEntityToReturn;
+  }
+
+  private buildBaseStepEntity(step: StepDto | (StepDto & { stepUuid: string })) {
+    return {
+      template: {
+        type: step.type,
+        name: step.name,
+        controls: step.controls,
+        content: '',
+      },
+      name: step.name,
+    };
+  }
+
+  private getPersistedStepIfFound(
+    persistedWorkflow: NotificationTemplateEntity | undefined,
+    stepUpdateRequest: StepUpdateDto | StepCreateDto
+  ) {
+    if (persistedWorkflow?.steps) {
```

Comment:
Prefer early null guards to reduce nesting in the code.

```suggestion
    if (!persistedWorkflow?.steps) {
      return 
    }
    
    // Then goes the for loop...
```